core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1739,18 +1756,24 @@ private[spark] class DAGScheduler(
           if (mapStage.rdd.isBarrier()) {
             // Mark all the map as broken in the map stage, to ensure retry all the tasks on
             // resubmitted stage attempt.
-            mapOutputTracker.unregisterAllMapOutput(shuffleId)
+            mapOutputTracker.unregisterAllMapAndMergeOutput(shuffleId)
           } else if (mapIndex != -1) {
             // Mark the map whose fetch failed as broken in the map stage
             mapOutputTracker.unregisterMapOutput(shuffleId, mapIndex, bmAddress)
+            if (mapStage.shuffleDep.shuffleMergeEnabled) {
```

Comment:
`MOT#unregisterMergeResult` can still be invoked if merge is disabled for the stage.
This if condition is only to avoid executing code unnecessarily?

Reply:
Instead of checking for shuffle dependency flag which can be set to false later only checking for the global `pushBasedShuffleEnabled` variable.